internal/topology/variables/patch_validation.go

```diff
@@ -0,0 +1,222 @@
+	var allErrs field.ErrorList
+	allErrs = append(allErrs,
+		validatePatchName(patch, names, path)...,
+	)
+	log.Print(allErrs.ToAggregate())
+	allErrs = append(allErrs,
+		validatePatchDefinitions(patch, variables, path)...,
+	)
+	return allErrs
+}
+
+func validatePatchName(patch clusterv1.ClusterClassPatch, names sets.String, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if patch.Name == "" {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				"patch names must not be empty",
+			),
+		)
+	}
+
+	if names.Has(patch.Name) {
+		allErrs = append(allErrs,
+			field.Invalid(
+				path.Child("name"),
+				patch.Name,
+				fmt.Sprintf("patch names must be unique. %s is repeated.", patch.Name),
+			),
+		)
+	}
+	return allErrs
+}
+
+/*
+
+
+- We consider validating the Go template by generating values for all variables used in the template, then evaluating
+the template and validating if the result is a valid YAML or JSON value.
+
+*/
+func validatePatchDefinitions(patch clusterv1.ClusterClassPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	for _, definition := range patch.Definitions {
+		allErrs = append(allErrs,
+			validatePatchDefinitionSelector(definition.Selector, path)...)
+
+		allErrs = append(allErrs,
+			validatePatchJSONPatches(definition.JSONPatches, variables, path)...)
+	}
+	return allErrs
+}
+
+func validatePatchJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	validOps := sets.NewString("add", "replace", "remove")
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("op").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch op must be one of \"add\" \"remove\" or \"replace\". %v is not valid.", jsonPatch.Op),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		if _, err := jsonptr.Parse(jsonPatch.Path); err != nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Child("path").Index(i),
+					jsonPatch,
+					fmt.Sprintf("jsonPatch path must be a valid JSON pointer (RFC 6901). %s not valid.", jsonPatch.Path),
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch must define one of Value or ValueFrom.",
```

Comment:
```suggestion
					"jsonPatch must define one of value or valueFrom.",
```
nit. Let's please use the spelling which the user would have to use (JSON path)
similar below